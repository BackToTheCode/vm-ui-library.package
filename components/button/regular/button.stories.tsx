import { checkA11y } from '@storybook/addon-a11y';
import { select, withKnobs } from '@storybook/addon-knobs';
import marked from 'marked';
import React from 'react';
import { Button } from '../regular';

export default {
  component: Button,
  decorators: [checkA11y],
  parameters: {
    info: {
      header: false,
      inline: true,

      text: marked(require('../button.stories.md').default)
    }
  },
  title: 'Design System|Buttons/Regular Button'
};

export const Primary = () => <Button variant={'primary'}>Go</Button>;

export const Secondary = () => <Button variant={'secondary'}>Go here</Button>;

export const Outline = () => <Button variant={'outline'}>Go back</Button>;

export const Text = () => <Button variant={'text'}>Click me</Button>;

export const WithKnobs = () => (
  <Button
    variant={select('Variant', ['primary', 'secondary', 'outline'], 'primary')}
  >
    Change me
  </Button>
);
WithKnobs.story = {
  decorators: [withKnobs]
};
