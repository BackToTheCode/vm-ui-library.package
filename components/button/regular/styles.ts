const styles = {
  button: {
    display: 'block',
    fontWeight: 'body',
    fontFamily: 'body',
    borderStyle: 'solid',
    borderWidth: 2,
    fontSize: 2,
    cursor: 'pointer',
    py: 5,
    px: 7,
    transition: '0.2s all',
    opacity: '1',
    filter: 'none',
    '&:hover': {
      transform: 'translateY(-1px)'
    },
    '&:focus': {
      outline: 0
    }
  },
  disabled: {
    cursor: 'default',
    opacity: '0.6',
    filter: 'grayscale(1)'
  }
};

export default styles;
