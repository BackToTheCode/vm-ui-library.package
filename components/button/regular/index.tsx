/** @jsx jsx */
import { jsx } from '@emotion/core';
import React, { FC } from 'react';
import { Button as BaseButton, SxStyleProp } from 'rebass';
import styles from './styles';

export interface ButtonProps {
  /** Click handler */
  onClick?: (e: React.MouseEvent<HTMLButtonElement, MouseEvent>) => void;

  /** Button variants */
  variant?: "primary" | "secondary" | "outline" | "text";

  isDisabled?: boolean;
  children?: React.ReactNode;

  /** System-UI style prop */
  sx?: SxStyleProp;
}

export const Button: FC<ButtonProps> = ({
  onClick,
  variant,
  isDisabled,
  sx,
  children
}) => {
  let mergedStyle = styles.button;
  if (isDisabled) {
    mergedStyle = { ...mergedStyle, ...styles.disabled };
  }

  return (
    <BaseButton
      onClick={onClick}
      sx={{ ...mergedStyle, ...sx }}
      variant={variant}
    >
      {children}
    </BaseButton>
  );
};

Button.defaultProps = {
  variant: 'primary'
};

Button.displayName = 'Button';