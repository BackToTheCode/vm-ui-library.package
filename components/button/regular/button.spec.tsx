import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { Button } from '../regular';
import { ThemeProvider } from 'emotion-theming';
import appTheme from '../../../styles/theme';

describe('<Button /> spec', () => {
  test('it renders the component', () => {
    // Arrange
    const buttonText = 'Click me';

    // Act
    const { getByText } = render(<Button>{buttonText}</Button>);

    // Assert
    expect(getByText(buttonText)).toMatchSnapshot();
  });

  test('that click gets handled', () => {
    // Arrange
    const buttonText = "Click me";
    jest.spyOn(window, 'alert').mockImplementation(() => {});
    const clickHandler = () => window.alert("Alerting you");
    
    // Act
    const { getByText } = render(<Button onClick={clickHandler}>{buttonText}</Button>);
    fireEvent.click(getByText(buttonText));

    // Assert
    expect(window.alert).toHaveBeenCalledTimes(1);
  });

  test('that <Button /> has correct background', () => {
    // Arrange
    const buttonText = 'Click me';
    const variant = "secondary";

    // Act 
    const { getByText } = render(<ThemeProvider theme={appTheme}><Button variant={variant}>{buttonText}</Button></ThemeProvider>);
    const button = getByText(buttonText);
    const style = window.getComputedStyle(button);

    // Assert
    expect(style.backgroundColor).toBe(appTheme.colors.secondary);
  });
});
