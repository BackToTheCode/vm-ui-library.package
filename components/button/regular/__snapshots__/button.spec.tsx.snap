// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> spec it renders the component 1`] = `
<button
  class="css-ao61va"
>
  Click me
</button>
`;
